gh-142598: Add new type of executable to support external JITs

DinoV opened the pull request:
https://github.com/python/cpython/pull/105727 introduced the ability to have different types of executables. During the discussion of that there was mention of potentially adding support for JITs to be able to plug in as well. There's a similar disucssion of the idea over here: https://github.com/faster-cpython/ideas/issues/575

This extends the existing set of executors so that PEP-523 JITs can hook into the interpreter frames and create their own "lightweight" frames which live on the stack with some minimal amount of initialization.

Currently the initialization required is probably more than idea:
* in order for frames to appear complete they need to eagerly initialize instr_ptr
* Some things assume f_funcobj is initialized
* Because the frames are currently `FRAME_OWNED_BY_THREAD` various things look at `stackpointer`.

Presumably the next step after this PR would be to add a new frame owner which is complete but doesn't require these additional fields to be initialized.

The JIT executors are created with a fixed object that includes a callback to reify the frame, a code object, and an optional piece of state. The reifier is called in various places where the JIT needs to populate the frame - for example accessing globals, builtins, or instr_ptr. Currently the API just provides a single callback with no information on what's required but you could imagine we could have a finer grained control if it was desired - but reification of the frame is generally going to happen in the slow path so simplicity seems the best here.

The code object isn't strictly necessary but because profilers and other external introspection tools frequently want a minimal number of hops to determine a functions identity it's tracked here instead of making them head back to the function object and then read the code from there.

<!-- gh-issue-number: gh-142598 -->
* Issue: gh-142598
<!-- /gh-issue-number -->

markshannon commented:
I think this needs its own issue, and a clearer specification. Not PEP level, but a specification in the issue would be good.

Some things the spec should cover:
* Should this be limited to PEP 523 JITs or should it be more general, so that the CPython JIT and any future versions of PyTorch dynamo can use it.
* JITs need to access fields of the frame, how does the API avoid infinite recursion when the reification needs part of the frame. E.g. it might need to access the function to get to the globals.
* Is the JIT expected to be able to execute a reified frame, or is it allowed to "unreify" the frame?
* Handling exceptions in during reification is tricky. Would it be better if the reification were infallible?

Some comments on this PR:
* Why special case the code object?
* Given that jitted frames won't have all the fields on a interpreter frame,  would it make sense for the stack to be of a different, smaller struct? Then the casting to the full frame could incorporate reification if necessary. [like this](https://github.com/python/cpython/compare/main...faster-cpython:cpython:push-extension-frame#diff-dc43b169f3202922a7787b97210ce299172f0cf03bd88529d473128f32e49fe4R65)
* Why check the type of the executable, rather than adding a new frame state? A frame state check would be faster.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

DinoV committed:
Add ability for external executor

DinoV commented:
@markshannon I've updated this to try and greatly simplify it. I've dropped the changes to the owner fields and gone back to using the type of the executable as we now check it a lot less. I've also made it so we also zero the instruction pointer and allow that as a complete frame. So now we're mostly only checking for external frames around instruction/line number operations and in the GC where it'd be good to update the stack for deferred reference counts. Let me know what you think!
